doc/hacking.rst

```diff
@@ -0,0 +1,200 @@
+
+.. code:: ocaml
+
+   let%expect_test "<test name>" =
+      print_string "Hello, world!";
+      [%expect {|
+        Hello, world!
+      |}]
+
+For integration tests, we use a system similar to `cram tests
+<https://bitheap.org/cram/>`_ for testing shell commands and their behavior:
+
+.. code:: bash
+
+   $ echo 'Hello, world!'
+   Hello, world!
+
+   $ false
+   [1]
+
+   $ cat <<EOF
+   > multi
+   > line
+   > EOF
+   multi
+   line
+
+.. _ppx_expect:      https://github.com/janestreet/ppx_expect
+
+Guidelines
+----------
+
+As with any long running software project, code written by one person will
+always eventually end up being maintained by another. Just like normal code, it
+is important to document tests. Especially since test suites are most often
+composed of many individual tests that must be understood on their own.
+
+A well written test case should be easy to understand. A reader should be able
+to quickly understand what property the test is checking, how it is doing it and
+how to convince one-self that the test outcome is the right one. A well written
+test will make it easy for future maintainers to understand the test and react
+when the test breaks. Most often, the code will need to be adapted to preserve
+the existing behavior, however in some rare cases the test expectation will need
+to be updated.
+
+It is crucial that each test cases makes it purpose and logic crystal clear so
+that future maintainers know how to deal with it.
+
+When writing a test, we generally have a good idea of what we want to test.
+Sometimes, we want to test that a new feature we developed is behaving as we
+expect. Sometimes, we want to add a reproduction case for a bug reported by a
+user to make sure future changes won't re-introduce the faulty behaviour. Just
+like when programming, we turn such an idea into code, which is a formal
+language that a computer can understand. While another person reading this code
+might be able to follow and understand what the code is doing step by step, it
+is not clear that they will be able to reconstruct the original idea the
+developer had in their mind when they originally wrote the code. What is worse,
+they might understand the code in a completely different way which would lead
+them to update it the wrong way.
+
+Adding Stanzas
+==============
+
+Adding new stanzas is the most natural way to extend dune with new features.
+Therefore we try to make this as easy as possible. The minimal amount of steps
+to add a new stanza is:
+
+- Extend ``Stanza.t`` with a new constructor to represent the new stanza
+- Modify ``Dune_file`` to parse the dune language into this constructor
+- Modify the rules  to interpret this stanza into rules. This is usually done in
+  ``Gen_rules```
+
+Versioning
+----------
+
+Dune is incredibly strict with versioning of new features, and modifications
+that are visible to the user. This means that any added stanza must be guarded
```

Comment:
maybe worth mentioning that _modifying_ existing rule behaviour also needs to be done carefully with version checks, not just adding new stanzas

Reply:
good idea.